_ext/eclipse-base/src/main/java/com/diffplug/spotless/extra/eclipse/base/SpotlessEclipseServiceConfig.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2016 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.extra.eclipse.base;
+
+import static com.diffplug.spotless.extra.eclipse.base.SpotlessEclipseFramework.LINE_DELIMITER;
+
+import java.util.Map;
+
+import org.eclipse.core.runtime.content.IContentTypeManager;
+import org.eclipse.core.runtime.preferences.IPreferencesService;
+import org.eclipse.osgi.service.datalocation.Location;
+import org.eclipse.osgi.service.debug.DebugOptions;
+import org.eclipse.osgi.service.environment.EnvironmentInfo;
+
+import com.diffplug.spotless.extra.eclipse.base.service.*;
+
+/**
+ * Configuration/Provision of services which shall be provided by the SpotlessEclipseFramework.
+ * <p>
+ * The SpotlessEclipseFramework plugins are customized by these services to provide the minimal
+ * functionality as required by the Spotless formatter.
+ * </p>
+ */
+public interface SpotlessEclipseServiceConfig {
+
+	/** Sets property/preference value available to all bundles, plugins and services. */
+	void set(String key, String value);
+
+	/** Sets property/preference values available to all bundles, plugins and services. */
+	default void set(Map<String, String> properties) {
+		properties.entrySet().stream().forEach(x -> set(x.getKey(), x.getValue()));
```

Comment:
Thanks for reminding me that Properties has its own forEach. I'll fix that.
But I think when you want to use local methods as consumers, you have to declare them. For example:
``` 
BiConsumer<String, String> setVal = (k,v) -> this.set(k,v);
properties.forEach(setVal);
```
Or have a I missed something?

Reply:
I think you're right in that in this case, rather than using `SpotlessEclipseServiceConfig::set`, the thing to use would be `this::set`, as we're referring to an instance method rather than a static method.

However, by my understanding (I haven't checked this in an IDE), it should be possible to write just this:

```java
properties.forEach(this::set);
```

as the signature of `this::set` matches `Map::forEach`.

But if that doesn't compile, then the following should work perfectly fine too:

```java
properties.forEach((k, v) -> set(k, v));
```

Reply:
Thanks @jbduncan for all your effort. I highly appreciate it. I know that it is not the most interesting thing to do. There were a few things I must admit, I should have noticed when I read my own comments. But some of the mistakes I made, I made already for many years. And since nobody told me that it's wrong...

Yes, I already made further changes, but not related to the issues you found. Most of it, was related to the 'miss-usage' of BundleController.SystemBundle, which was no longer a system bundle how OSGi defines it. And I needed an extension/configuration to run different bundles in different states, depending who asks. 
My changes, probably not flawless, improved the readability code, so I think its OK to commit them without further review. Thanks again for your work.